// ora-halo/index.js
var Mesh        = require('pex-glu').Mesh;
var MeshExt     = require('../glu/MeshExt');
var Geometry    = require('pex-geom').Geometry;
var Vec3        = require('pex-geom').Vec3;
var Vec2        = require('pex-geom').Vec2;
var Color       = require('pex-color').Color;
var Texture2D   = require('pex-glu').Texture2D;
var makeCircle  = require('make-circle');
var HaloRingMaterial = require('ora-halo-ring-material');
var remap       = require('re-map');
var Time        = require('pex-sys').Time;
var glu         = require('pex-glu');
var floor       = Math.floor;
var pow         = Math.pow;
var R           = require('ramda');
var TextBox     = require('../typo/TextBox');

function Halo(opts) {
  this.minRingRadius = 0.2;
  this.maxRingRadius = 1.0;
  this.minNumRings = 0;
  this.maxNumRings = 80;
  this.size = 0;
  this.speed = 0.5;
  this.color = 0;
  this.colorCenter = 0;
  this.colorCenterRatio = 0;
  this.complexity = 0;
  this.brightness = 1;
  this.wobble = 0;
  this.time = 0;
  this.mode = 'present';
  this.enableLabels = false;
  this.background = new Color(0,0,0,1);
  this.backgroundSolid = new Color(0,0,0,0);
  this.grid = new Color(1,0,0,1);
  this.gradient = [Color.Red, Color.Blue];
  this.growth = 0.05;
  this.glow = 0.75;
  this.scale = 40;

  this.ringInstances = [];
  this.buildMesh(opts);

  this.gridLineInstances = [];
  this.gridLineLabels = [];
  this.updateGridLines();
}

Halo.prototype.buildMesh = function(opts) {
  var lineDotsTexture = this.lineDotsTexture = Texture2D.load(opts.lineDotsTexture, { repeat: true, mipmap: true });
  var lineSolidTexture = this.lineSolidTexture = Texture2D.load(opts.lineSolidTexture, { repeat: true, mipmap: true });
  var colorTexture = this.colorTexture = Texture2D.load(opts.colorTexture, { repeat: false, mipmap: true });
  this.material = new HaloRingMaterial({ lineTexture: lineDotsTexture, colorTexture: colorTexture } );
  var gridColorTexture = this.gridColorTexture = Texture2D.load(opts.gridColorTexture, { repeat: false, mipmap: true });
  this.gridMaterial = new HaloRingMaterial({ lineTexture: lineDotsTexture, colorTexture: gridColorTexture } );

  this.updateGeometry(this.mode);
}

Halo.prototype.setMode = function(mode) {
  this.mode = mode;
  this.ringInstances = [];
  this.updateGeometry(this.mode);
}

Halo.prototype.updateGeometry = function(mode) {
  var numPoints = (mode == 'timeline') ? 32 : 128;
  var points    = makeCircle(1, numPoints, 'x', 'z', true);
  var texCoords = points.map(function(v, i) { return new Vec2(i/(points.length-1), 0)});
  var edges     = points.map(function(v, i) { return [ i, (i+1) % points.length ]});
  edges.pop();

  var g = new Geometry({ vertices: points, texCoords: texCoords, edges: edges  });

  if (this.ringMesh) {
    this.ringMesh.dispose();
    this.gridMesh.dispose();
  }
  this.ringMesh = new Mesh(g, this.material, { lines: true });
  this.gridMesh = new Mesh(g, this.gridMaterial, { lines: true });
}

Halo.prototype.updateGridLines = function() {
  var numGridLines = 10;
  for(var i=1; i<=numGridLines; i++) {
    var k = i/numGridLines;
    var radius = remap(k*k, 0, 1, this.minRingRadius, this.maxRingRadius*1.1);
    this.gridLineInstances.push({
      uniforms: {
        radius: radius,
        color: 0,
        colorCenter: 0,
        colorCenterRatio: 0,
        complexity: 0,
        speed: 0.5,
        time: 0,
        alpha: 0.5,
        brightness: 1,
        wobble: 0,
        wobbleFrequency: 0.08,
        horizontalOffset: 0,
        horizontalOffset2: 2.5,
        stratified: false,
      }
    })
    if (i >= 2 && this.enableLabels) {
      var r = '' + i/10;
      if (r == '1') r = '1.0';
      var label = new TextBox('' + r, 'Arial', 10, { origin: TextBox.Origin.Left });
      label.radius = radius;
      this.gridLineLabels.push(label);
    }
  }
}

Halo.prototype.setGlobalParam = function(param, value) {
  if (this[param] !== undefined && typeof(this[param]) !== 'function') {

    if (param == 'background') {
        if (value instanceof Color) {
            this[param] = value;
        }
        else {
            this[param] = Color.fromHex(value);
        }
    }
    else {
        this[param] = value;
    }
  }
}

Halo.prototype.update = function() {
  if (this.mode == 'present') {
    this.updateParams();
  }
  this.animate();
}

Halo.prototype.updateParams = function() {
  var numRings = floor(remap(this.size, 0, 1, this.minNumRings, this.maxNumRings));

  var lastRing = this.ringInstances[this.ringInstances.length-1];

  var speed = remap(this.speed, 0.0, 1.0, 0.2, 1.0);

  this.time += speed * Time.delta;

  while(this.ringInstances.length < numRings) {
    //$("#sizeSlider").slider({ change: function(event, ui) { HaloSetParam('size', ui.value/100); } })
    //$("#colorSlider").slider({ change: function(event, ui) { HaloSetParam('color', ui.value/100); } })
    //$("#complexitySlider").slider({ change: function(event, ui) { HaloSetParam('complexity', ui.value/100); } })
    //$("#speedSlider").slider({ change: function(event, ui) { HaloSetParam('speed', ui.value/100); } })
    //$("#brightnessSlider").slider({ change: function(event, ui) { HaloSetParam('brightness', ui.value/100); } })
    //$("#wobbleSlider").slider({ change: function(event, ui) { HaloSetParam('wobble', ui.value/100); } })

    this.ringInstances.push({
      uniforms: {
        radius: 1,
        color: 0,
        colorCenter: 0,
        colorCenterRatio: 0,
        complexity: 0,
        speed: 0.5,
        time: lastRing ? lastRing.uniforms.time : 0,
        alpha: 0,
        brightness: 1,
        wobble: 0,
        wobbleFrequency: 0.08,
        horizontalOffset: 0,
        horizontalOffset2: 2.5,
        stratified: false
      },
      uniformsTargets: {
        radius: 1,
        color: 0,
        colorCenter: 0,
        colorCenterRatio: 0,
        complexity: 0,
        speed: 0.5,
        time: 0,
        alpha: 1,
        brightness: 1,
        wobble: 0,
        wobbleFrequency: 0.08,
        horizontalOffset: 0,
        horizontalOffset2: 2.5,
        stratified: false
      },
    })
  }

  for(var i=0; i<this.ringInstances.length; i++) {
    var ring = this.ringInstances[i];
    ring.uniformsTargets.alpha = (i < numRings) ? 1 : 0;

    var k = remap(i, 0, this.maxNumRings, 0, 1);
    ring.uniformsTargets.radius = remap(k*k, 0, 1, this.minRingRadius, this.maxRingRadius);
    ring.uniforms.radius = remap(k*k, 0, 1, this.minRingRadius, this.maxRingRadius);

    ring.uniformsTargets.color = this.color;
    ring.uniformsTargets.colorCenter = this.colorCenter;
    ring.uniformsTargets.colorCenterRatio = this.colorCenterRatio;
    ring.uniformsTargets.complexity = this.complexity;
    ring.uniformsTargets.speed = remap(this.speed, 0.0, 1.0, 0.2, 1.0);
    ring.uniformsTargets.brightness = remap(this.brightness, 0.0, 1.0, 0.3, 1.0);
    ring.uniformsTargets.wobble = remap(this.wobble, 0.0, 1.0, 0.0, 1.0);
  }
}

Halo.prototype.addTimeStamp = function(params) {
  console.log('addTimeStamp', params.complexity)
  var lastRing = this.ringInstances[this.ringInstances.length-1];
  this.ringInstances.push({
    uniforms: {
      radius: 1,
      color: params.color,
      colorCenter: params.colorCenter,
      colorCenterRatio: params.colorCenterRatio,
      complexity: params.complexity,
      speed: params.speed,
      time: 0,//lastRing ? lastRing.uniforms.time : 0,
      alpha: 0,
      brightness: 1,
      wobble: 0,
      radiusHeightScale: 1,
      radiusHeightLocalScale: 1,
      horizontalOffset: 0,
      horizontalOffset2: 2.5,
      stratified: true
    },
    uniformsTargets: {
      radius: 1,
      color: params.color,
      colorCenter: params.colorCenter,
      colorCenterRatio: params.colorCenterRatio,
      complexity: params.complexity,
      speed: params.speed,
      time: 0,
      alpha: 1,
      brightness: 1,
      wobble: 0,
      radiusHeightScale: 1,
      radiusHeightLocalScale: 1,
      horizontalOffset: 0,
      horizontalOffset2: 0,
      stratified: true
    },
  })

  if (this.ringInstances.length > this.maxNumRings) {
    this.ringInstances.shift();
  }

  for(var i=0; i<this.ringInstances.length; i++) {
    var ring = this.ringInstances[i];
    var k = remap(i, 0, this.maxNumRings, 0, 1);
    ring.uniformsTargets.radius = remap(k*k, 0, 1, this.minRingRadius, this.maxRingRadius);
  }

  //if (this.rings)
}

Halo.prototype.animate = function() {
  for(var i=0; i<this.ringInstances.length; i++) {
    var ring = this.ringInstances[i];

    ring.uniforms.radiusHeightScale = remap(i, 0, this.maxNumRings, 0, 1);
    ring.uniforms.radiusHeightLocalScale = remap(i, 0, this.ringInstances.length, 0, 1);
    ring.uniforms.radius      += (ring.uniformsTargets.radius - ring.uniforms.radius) * this.growth;
    ring.uniforms.color       += (ring.uniformsTargets.color - ring.uniforms.color) * this.growth;
    ring.uniforms.colorCenter      += (ring.uniformsTargets.colorCenter - ring.uniforms.colorCenter) * this.growth;
    ring.uniforms.colorCenterRatio += (ring.uniformsTargets.colorCenterRatio - ring.uniforms.colorCenterRatio) * this.growth;
    ring.uniforms.complexity  += (ring.uniformsTargets.complexity - ring.uniforms.complexity) * this.growth;
    ring.uniforms.speed       += (ring.uniformsTargets.speed - ring.uniforms.speed) * this.growth;
    ring.uniforms.brightness  += (ring.uniformsTargets.brightness - ring.uniforms.brightness) * this.growth;
    ring.uniforms.alpha       += (ring.uniformsTargets.alpha - ring.uniforms.alpha) * this.growth/2;
    ring.uniforms.wobble      += (ring.uniformsTargets.wobble - ring.uniforms.wobble) * this.growth/2;
    ring.uniforms.index       = i/this.ringInstances.length;
    ring.uniforms.time        = this.time;

    ring.uniforms.horizontalOffset =  5 * (Time.seconds/5 - Math.floor(Time.seconds/5));
    ring.uniforms.horizontalOffset2  = 5 * ((Time.seconds + 2.5)/5 - Math.floor((Time.seconds + 2.5)/5));

    if (ring.uniformsTargets.alpha == 0 && ring.uniforms.alpha < 0.01) {
      this.ringInstances.splice(i, 1);
      i--;
    }
  }
}

Halo.prototype.setColorTexture = function(path) {
  this.colorTexture.dispose();
  this.colorTexture = Texture2D.load(path);
  this.ringMesh.material.uniforms.colorTexture = this.colorTexture;
  this.ringMesh.drawInstancesOpt(camera, this.gridLineInstances);
}

Halo.prototype.draw = function(camera, camera2D, w, h) {
  if (!this.ringMesh.material.program.ready) return;
  glu.clearColor(Color.Transparent);
  glu.enableDepthReadAndWrite(false, false);
  glu.enableAdditiveBlending(true);
  this.ringMesh.material.uniforms.lineTexture = this.lineDotsTexture;
  this.ringMesh.drawInstancesOpt(camera, this.ringInstances);
  this.gridMesh.drawInstancesOpt(camera, this.gridLineInstances);
  this.gridLineLabels.forEach(function(label, i) {
    var pos = camera.getScreenPos({ x: 0, y: 0, z:  label.radius * 2}, w, h);
    var pos3 = label._position;
    pos3.copy(pos);
    label.setPosition(pos3);
    label.draw(camera2D);
  }.bind(this));
  glu.enableAlphaBlending(false, false);
}

Halo.prototype.drawSolid = function(camera) {
  this.backgroundSolid.r = this.background.r;
  this.backgroundSolid.g = this.background.g;
  this.backgroundSolid.b = this.background.b;
  glu.clearColor(this.backgroundSolid);
  if (!this.ringMesh.material.program.ready) return;
  this.ringMesh.material.uniforms.lineTexture = this.lineSolidTexture;
  this.ringMesh.drawInstancesOpt(camera, this.ringInstances);
}

Halo.prototype.dispose = function() {
  this.mesh.material.dispose();
  this.mesh.dispose();
}

module.exports = Halo;
