// ora-halo/index.js
var Vec3        = require('pex-math/Vec3');
var Vec2        = require('pex-math/Vec2');
var Color       = require('pex-color');
var makeCircle  = require('make-circle');
var remap       = require('re-map');
var R           = require('ramda');
//TODO: var TextBox     = require('../typo/TextBox');
var createGradient = require('create-gradient');
var isBrowser   = require('is-browser');
var io          = require('pex-io');
var glslify     = require('glslify-sync');

var floor       = Math.floor;
var pow         = Math.pow;

var HALO_RING_VERT = glslify(__dirname + '/assets/HaloRing.vert')
var HALO_RING_FRAG = glslify(__dirname + '/assets/HaloRing.frag')

var HaloModes = {
    PRESENT: 'present',
    TIMELINE: 'timeline'
}

if (isBrowser) {
    io.loadImage = function(url, callback) {
        var img = new Image();
        img.onload = function() {
            callback(null, img);
        }
        img.onerror = function(err) {
            console.log('err', err);
        }
        img.src = url;
    }
}


function Halo(ctx, window, opts) {
  this.ctx = ctx;
  this.window = window;
  this.minRingRadius = 0.2;
  this.maxRingRadius = 1.0;
  this.minNumRings = 0;
  this.maxNumRings = 80;
  this.size = 0;
  this.speed = 0.5;
  this.color = 0;
  this.colorCenter = 0;
  this.colorCenterRatio = 0;
  this.complexity = 0;
  this.brightness = 1;
  this.wobble = 0;
  this.time = 0;
  this.mode = HaloModes.PRESENT;
  this.enableLabels = false;
  this.background = [0,0,0,1];
  this.backgroundTransparent = [0,0,0,0];
  this.grid = [0.75, 0.75, 0.75, 1];
  this.showGrid = true;
  this.spectrum = [[0.1,0.1,0.1,1]];
  this.growth = 0.05;
  this.glow = 0.75;
  this.scale = 40;
  this.ringResolution = 128; //32
  //if true then all rings at the same angle / clock hour will have the same freequency sync
  this.stratified = false;

  this.ringInstances = [];
  this.buildMesh(opts);

  this.gridLineInstances = [];
  this.gridLineLabels = [];
  this.updateGridLines();

  this.solidLines = false;
  this.evenLineDistribution = false;
}

Halo.prototype.buildMesh = function(opts) {
  var ctx = this.ctx;

  var lineDotsTexture = this.lineDotsTexture = ctx.createTexture2D(null, 256, 32, { repeat: true, mipmap: true });
  io.loadImage(opts.lineDotsTexture, function(err, img) {
    lineDotsTexture.update(img, img.width, img.height, { repeat: true, mipmap: true });
    ctx.bindTexture(lineDotsTexture);
  }.bind(this))

  var lineSolidTexture = this.lineSolidTexture = ctx.createTexture2D(null, 256, 32, { repeat: true, mipmap: true });
  io.loadImage(opts.lineSolidTexture, function(err, img) {
    lineSolidTexture.update(img, img.width, img.height, { repeat: true, mipmap: true });
    ctx.bindTexture(lineDotsTexture);
  }.bind(this))

  var colorTexture = this.colorTexture = ctx.createTexture2D(null, 256, 32);
  io.loadImage(opts.colorTexture, function(err, img) {
    colorTexture.update(img, img.width, img.height, { repeat: false });
  })
  var colorSpectrumTexture = this.colorSpectrumTexture = ctx.createTexture2D(createGradient(this.spectrum, 256, 32), 256, 32);
  var gridColorTexture = this.gridColorTexture = ctx.createTexture2D(createGradient([this.grid], 256, 32), 256, 32);

  this.ringProgram = ctx.createProgram(HALO_RING_VERT, HALO_RING_FRAG);
  this.gridProgram = ctx.createProgram(HALO_RING_VERT, HALO_RING_FRAG);

  this.defaultRingUniforms = { lineTexture: lineDotsTexture, colorTexture: colorTexture }
  this.defaultGridUniforms = { lineTexture: lineDotsTexture, colorTexture: gridColorTexture };

  this.updateGeometry(this.mode);
}

Halo.prototype.setMode = function(mode) {
  this.mode = mode;
  this.ringInstances = [];
  this.updateGeometry(this.mode);
}

Halo.prototype.updateGeometry = function(mode) {
  var ctx = this.ctx;
  var numPoints = this.ringResolution;
  var points    = makeCircle(1, numPoints, 0, 2, true); //x,z plane
  var texCoords = points.map(function(v, i) { return [i/(points.length-1), 0]});
  var edges     = points.map(function(v, i) { return [ i, (i+1) % points.length ]});
  edges.pop();

  var attributes = [
    { data: points, location: ctx.ATTRIB_POSITION },
    { data: texCoords, location: ctx.ATTRIB_TEX_COORD_0 }
  ];
  var indices = { data: edges, usage: ctx.STATIC_DRAW };

  if (this.ringMesh) {
    //TODO: this.ringMesh.dispose();
    //TODO: this.gridMesh.dispose();
  }
  this.ringMesh = ctx.createMesh(attributes, indices, ctx.LINES);
  this.gridMesh = ctx.createMesh(attributes, indices, ctx.LINES);
}

Halo.prototype.updateGridLines = function() {
  this.gridLineInstances = [];
  var numGridLines = 10;
  for(var i=1; i<=numGridLines; i++) {
    var k = i/numGridLines;
    var k2 = this.evenLineDistribution ? k : k * k;
    var radius = remap(k2, 0, 1, this.minRingRadius, this.maxRingRadius*1.1);
    this.gridLineInstances.push({
      uniforms: {
        radius: radius,
        color: 0,
        colorCenter: 0,
        colorCenterRatio: 0,
        complexity: 0,
        speed: 0.5,
        time: 0,
        alpha: 0.5,
        brightness: 1,
        wobble: 0,
        wobbleFrequency: 0.08,
        horizontalOffset: 0,
        horizontalOffset2: 2.5,
        stratified: this.stratified,
      }
    })
    if (i >= 2 && this.enableLabels) {
      var r = '' + i/10;
      if (r == '1') r = '1.0';
      var label = new TextBox('' + r, 'Arial', 10, { origin: TextBox.Origin.Left });
      label.radius = radius;
      this.gridLineLabels[i] = label;
    }
  }
}

Halo.prototype.setGlobalParam = function(param, value) {
  if (this[param] !== undefined && typeof(this[param]) !== 'function') {

    if (param == 'background') {
        if (Array.isArray(value)) {
            this[param] = value;
        }
        else {
            this[param] = Color.fromHex(value);
        }
    }
    else if (param == 'spectrum') {
        this.setColorSpectrum(value);
    }
    else {
        this[param] = value;
    }

    if (param == 'minRingRadius') {
        this.updateGridLines();
    }
    if (param == 'ringResolution') {
        this.updateGeometry();
    }
  }
}

Halo.prototype.update = function() {
  if (this.mode == HaloModes.PRESENT) {
    this.updateParams();
  }
  this.animate();
}

Halo.prototype.updateParams = function() {
  var numRings = floor(remap(this.size, 0, 1, this.minNumRings, this.maxNumRings));

  var lastRing = this.ringInstances[this.ringInstances.length-1];

  var speed = remap(this.speed, 0.0, 1.0, 0.2, 1.0);

  this.time += speed * this.window.getTime().getDeltaSeconds();

  while(this.ringInstances.length < numRings) {
    //$("#sizeSlider").slider({ change: function(event, ui) { HaloSetParam('size', ui.value/100); } })
    //$("#colorSlider").slider({ change: function(event, ui) { HaloSetParam('color', ui.value/100); } })
    //$("#complexitySlider").slider({ change: function(event, ui) { HaloSetParam('complexity', ui.value/100); } })
    //$("#speedSlider").slider({ change: function(event, ui) { HaloSetParam('speed', ui.value/100); } })
    //$("#brightnessSlider").slider({ change: function(event, ui) { HaloSetParam('brightness', ui.value/100); } })
    //$("#wobbleSlider").slider({ change: function(event, ui) { HaloSetParam('wobble', ui.value/100); } })

    this.ringInstances.push({
      uniforms: {
        radius: 1,
        color: 0,
        colorCenter: 0,
        colorCenterRatio: 0,
        complexity: 0,
        speed: 0.5,
        time: lastRing ? lastRing.uniforms.time : 0,
        alpha: 0,
        brightness: 1,
        wobble: 0,
        wobbleFrequency: 0.08,
        horizontalOffset: 0,
        horizontalOffset2: 2.5,
        stratified: this.stratified
      },
      uniformsTargets: {
        radius: 1,
        color: 0,
        colorCenter: 0,
        colorCenterRatio: 0,
        complexity: 0,
        speed: 0.5,
        time: 0,
        alpha: 1,
        brightness: 1,
        wobble: 0,
        wobbleFrequency: 0.08,
        horizontalOffset: 0,
        horizontalOffset2: 2.5,
        stratified: this.stratified
      },
    })
  }

  while(this.ringInstances.length > 0 && this.ringInstances.length > this.maxNumRings) {
      this.ringInstances.pop();
  }

  for(var i=0; i<this.ringInstances.length; i++) {
    var ring = this.ringInstances[i];
    ring.uniformsTargets.alpha = (i < numRings) ? 1 : 0;

    var k = remap(i, 0, this.maxNumRings, 0, 1);
    var k2 = this.evenLineDistribution ? k : k * k;
    ring.uniformsTargets.radius = remap(k2, 0, 1, this.minRingRadius, this.maxRingRadius);
    ring.uniforms.radius = remap(k2, 0, 1, this.minRingRadius, this.maxRingRadius);

    ring.uniformsTargets.color = this.color;
    ring.uniformsTargets.colorCenter = this.colorCenter;
    ring.uniformsTargets.colorCenterRatio = this.colorCenterRatio;
    ring.uniformsTargets.complexity = this.complexity;
    ring.uniformsTargets.speed = remap(this.speed, 0.0, 1.0, 0.2, 1.0);
    ring.uniformsTargets.brightness = remap(this.brightness, 0.0, 1.0, 0.3, 1.0);
    ring.uniformsTargets.wobble = remap(this.wobble, 0.0, 1.0, 0.0, 1.0);
  }
}

Halo.prototype.addTimeStamp = function(params) {
  console.log('addTimeStamp', params.complexity)
  var lastRing = this.ringInstances[this.ringInstances.length-1];
  this.ringInstances.push({
    uniforms: {
      radius: 1,
      color: params.color,
      colorCenter: params.colorCenter,
      colorCenterRatio: params.colorCenterRatio,
      complexity: params.complexity,
      speed: params.speed,
      time: 0,//lastRing ? lastRing.uniforms.time : 0,
      alpha: 0,
      brightness: 1,
      wobble: 0,
      radiusHeightScale: 1,
      radiusHeightLocalScale: 1,
      horizontalOffset: 0,
      horizontalOffset2: 2.5,
      stratified: this.stratified
    },
    uniformsTargets: {
      radius: 1,
      color: params.color,
      colorCenter: params.colorCenter,
      colorCenterRatio: params.colorCenterRatio,
      complexity: params.complexity,
      speed: params.speed,
      time: 0,
      alpha: 1,
      brightness: 1,
      wobble: 0,
      radiusHeightScale: 1,
      radiusHeightLocalScale: 1,
      horizontalOffset: 0,
      horizontalOffset2: 0,
      stratified: this.stratified
    },
  })

  for(var i=0; i<this.ringInstances.length - this.maxNumRings; i++) {
      //this.ringInstances.shift();
      this.ringInstances[i].uniformsTargets.alpha = 0;
  }

  for(var i=0; i<this.ringInstances.length; i++) {
    var ring = this.ringInstances[i];
    var k = remap(i, 0, this.maxNumRings, 0, 1);
    var k2 = this.evenLineDistribution ? k : k * k;
    ring.uniformsTargets.radius = remap(k2, 0, 1, this.minRingRadius, this.maxRingRadius);
  }
}

Halo.prototype.animate = function() {
    var needRadiusUpdate = false;
  for(var i=0; i<this.ringInstances.length; i++) {
    var ring = this.ringInstances[i];

    ring.uniforms.radiusHeightScale = remap(i, 0, this.maxNumRings, 0, 1);
    ring.uniforms.radiusHeightLocalScale = remap(i, 0, this.ringInstances.length, 0, 1);
    ring.uniforms.radius      += (ring.uniformsTargets.radius - ring.uniforms.radius) * this.growth;
    ring.uniforms.color       += (ring.uniformsTargets.color - ring.uniforms.color) * this.growth;
    ring.uniforms.colorCenter      += (ring.uniformsTargets.colorCenter - ring.uniforms.colorCenter) * this.growth;
    ring.uniforms.colorCenterRatio += (ring.uniformsTargets.colorCenterRatio - ring.uniforms.colorCenterRatio) * this.growth;
    ring.uniforms.complexity  += (ring.uniformsTargets.complexity - ring.uniforms.complexity) * this.growth;
    ring.uniforms.speed       += (ring.uniformsTargets.speed - ring.uniforms.speed) * this.growth;
    ring.uniforms.brightness  += (ring.uniformsTargets.brightness - ring.uniforms.brightness) * this.growth;
    ring.uniforms.alpha       += (ring.uniformsTargets.alpha - ring.uniforms.alpha) * this.growth/2;
    ring.uniforms.wobble      += (ring.uniformsTargets.wobble - ring.uniforms.wobble) * this.growth/2;
    ring.uniforms.index       = i/this.ringInstances.length;
    ring.uniforms.time        = this.time;

    var seconds = this.window.getTime().getElapsedSeconds();

    ring.uniforms.horizontalOffset =  5 * (seconds/5 - Math.floor(seconds/5));
    ring.uniforms.horizontalOffset2  = 5 * ((seconds + 2.5)/5 - Math.floor((seconds + 2.5)/5));


    if (ring.uniformsTargets.alpha == 0 && ring.uniforms.alpha < 0.01) {
      this.ringInstances.splice(i, 1);
      i--;
      needRadiusUpdate = true;
    }
  }

  if (needRadiusUpdate) {
      for(var i=0; i<this.ringInstances.length; i++) {
        var ring = this.ringInstances[i];
        var k = remap(i, 0, this.maxNumRings, 0, 1);
        var k2 = this.evenLineDistribution ? k : k * k;
        ring.uniformsTargets.radius = remap(k2, 0, 1, this.minRingRadius, this.maxRingRadius);
      }
  }
}

Halo.prototype.setColorTexture = function(path) {
  console.log('setColorTexture', setColorTexture);
  //TODO: this.colorTexture.dispose();
  //TODO: this.colorTexture = Texture2D.load(path);
  //TODO: this.ringMesh.material.uniforms.colorTexture = this.colorTexture;
  //TODO: this.ringMesh = new Mesh(g, this.material, { lines: true });
  //TODO: this.ringMesh.drawInstancesOpt(camera, this.gridLineInstances);
}

Halo.prototype.setColorSpectrum = function(colors) {
    this.spectrum = colors;
    var tex = this.colorSpectrumTexture;
    console.log('Halo.setColorSpectrum', colors, tex.getWidth(), tex.getHeight());

    var gradient = createGradient(colors, tex.getWidth(), tex.getHeight());
    tex.update(gradient);
    this.defaultRingUniforms.colorTexture = this.colorSpectrumTexture
}

Halo.prototype.drawMeshInstances = function(ctx, camera, program, instances, defaultUniforms) {
  ctx.bindProgram(program);
  ctx.bindMesh(this.ringMesh);

  var numTextures = 0;
  for (uniformName in defaultUniforms) {
      if (program.hasUniform(uniformName)) {
          ctx.bindTexture(defaultUniforms[uniformName], numTextures)
          program.setUniform(uniformName, numTextures);
          numTextures++;
      }
  }

  for (var i = 0; i < instances.length; i++) {
    var instance = instances[i];
    for (uniformName in instance.uniforms) {
        if (program.hasUniform(uniformName)) {
            program.setUniform(uniformName, instance.uniforms[uniformName])
        }
    }
    ctx.drawMesh();
  }
}

Halo.prototype.draw = function(camera, camera2D, w, h) {
  var ctx = this.ctx;

  this.defaultRingUniforms.lineTexture = this.solidLines ? this.lineSolidTexture : this.lineDotsTexture;

  ctx.pushState(ctx.LINE_WIDTH_BIT);
  ctx.setLineWidth(this.solidLines ? 1 : 2);
  this.drawMeshInstances(ctx, camera, this.ringProgram, this.ringInstances, this.defaultRingUniforms)
  ctx.setLineWidth(2);
  if (this.showGrid) {
      this.drawMeshInstances(ctx, camera, this.gridProgram, this.gridLineInstances, this.defaultGridUniforms)
  }
  ctx.popState();

  //TODO: this.gridLineLabels.forEach(function(label, i) {
    //var pos = camera.getScreenPos({ x: 0, y: 0, z:  label.radius * 2}, w, h);
    //var pos3 = label._position;
    //pos3.copy(pos);
    //label.setPosition(pos3);
    //label.draw(camera2D);
  //}.bind(this));
}

Halo.prototype.drawSolid = function(camera) {
  var ctx = this.ctx;
  this.backgroundTransparent[0] = this.background[0];
  this.backgroundTransparent[1] = this.background[1];
  this.backgroundTransparent[2] = this.background[2];
  ctx.pushState(ctx.LINE_WIDTH_BIT);
  ctx.setLineWidth(3);
  this.defaultRingUniforms.lineTexture = this.lineSolidTexture;
  this.drawMeshInstances(ctx, camera, this.ringProgram, this.ringInstances, this.defaultRingUniforms);
  ctx.popState();
}

Halo.prototype.dispose = function() {
  this.mesh.material.dispose();
  this.mesh.dispose();
}

Halo.HaloModes = HaloModes;

module.exports = Halo;
